### PR TITLE
Update and Expand the Graph API Tutorial

### LINKED ISSUE #220: Update DGraph Documentation

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #220

### CODE REVIEW

File: docs/tutorials/dgraph_tutorial.md [hunk 1]
```diff
@@ -1,99 +1,162 @@
...
-## Construct `DGraph` from TGB Datasets
+## 1. The Core Objects
 
-The [Temporal Graph Benchmark (TGB)](https://tgb.complexdatalab.com/) provides a suite of temporal graph datasets with diverse scales, properties, and tasks.
+TGM's graph API revolves around four main objects:
 
-TGM supports constructing `DGraph` objects from TGB [LinkPropPrediction](https://tgb.complexdatalab.com/docs/linkprop/) and [NodePropPrediction](https://tgb.complexdatalab.com/docs/nodeprop/) datasets. Temporal knowledge graphs (TKG) and temporal hypergraphs (THG) are not yet supported.
+| Object      | Description                                                   | Mutable | Device Semantics | Typical Usage                                    |
+| ----------- | ------------------------------------------------------------- | ------- | ---------------- | ------------------------------------------------ |
+| `DGData`    | Mutable bulk dataset storage (IO, splits, transforms)         | Yes     | No               | Ingesting datasets from disk, TGB, preprocessing |
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:20]: `DGData` is also essentially a group of tensors, maybe we should mention here

File: docs/tutorials/dgraph_tutorial.md [hunk 2]
```diff
@@ -1,99 +1,162 @@
...
+## 3. Constructing DGData
 
-## Create a Custom `DGraph`
+You can build datasets in multiple ways. Let's look at each.
 
-You can also define a `DGraph` from your own data.
+### 3.1 From TGB
 
-### Construct `DGraph` From Raw Tensors
+This is most likely all you need. The [Temporal Graph Benchmark (TGB)](https://tgb.complexdatalab.com/) provides a suite of temporal graph datasets with diverse scales and properties. We natively support direct construction from all the `tgbl-` and `tgbn-`.
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md]: instead of `This is most likely all you need` (which I like btw). instead we say `Here is how to use natively supported TGB data in TGM`.

File: docs/tutorials/dgraph_tutorial.md [hunk 3]
```diff
@@ -1,99 +1,162 @@
...
 
-#### Define Temporal Edges
+> **Note**: Temporal knowledge graph (TKG) and temporal hypergraph (THG) are not yet supported in TGM.
 
-- `edge_index`: shape `[num_edge_events, 2]`
-- `edge_timestamps`: shape `[num_edge_events]`
-- `edge_feats`: shape `[num_edge_events, D_edge]` (optional)
+> **Node**: To load a TGB dataset, you must have the `py-tgb` package in your python env.
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md]: or install TGB by source?

File: docs/tutorials/dgraph_tutorial.md [hunk 4]
```diff
@@ -1,99 +1,162 @@
...
 
 ```python
-import torch
+from tgm import DGData
 
-edge_index = torch.LongTensor([[2, 2], [2, 4], [1, 8]])
-edge_timestamps = torch.LongTensor([1, 5, 20])
-edge_feats = torch.rand(3, 5)  # optional edge features
+# Load the Wikipedia dataset from TGB
+data = DGData.from_tgb('tgbl-wiki')
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:91]: idea: maybe we can have a print function which prints out a table showing what is available and what is not available for native datasets in `DGData`

Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md:91]: actually it would apply to all `DGData`

File: docs/tutorials/dgraph_tutorial.md [hunk 5]
```diff
@@ -111,25 +174,151 @@ dg = DGraph.from_pandas(
...
+
+> **Note**: This is only well defined if the DGData time delta is *time-ordered*. If you try discretizing an event-ordered dataset, you will get a `tgm.exceptions.EventOrderedConversionError`.
+
+> **Note**: Discretization goes from finer time units (e.g. seconds) to coarse time units (e.g. hours). If your attempt to discretize in the other direction, you'll get a `tgm.exceptions.InvalidDiscretizationError`.
+
+See our [time management tutorial](../tutorials/time_delta_tutorial.md) for more details on discretization and how it relates to `TimeDeltaDG`.
+
+## 6. From `DGData` to `DGraph`
+
+Once your dataset is ready to go, you can cast it to `DGraph`:
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:301]: we should quickly mention that both edge and node events will be chronologically sorted therefore, for a user, imagine a stream of both edge and node events arriving in chronologically order

File: docs/tutorials/dgraph_tutorial.md [hunk 6]
```diff
@@ -111,25 +174,151 @@ dg = DGraph.from_pandas(
...
+- `tgm.exceptions.InvalidNodeIDError`: Raised when you dataset contains `-1` as a node ID (reserved for padding).
 
-`DGraph` objects act as views over the underlying data. You can access properties and perform slicing operations.
+## 4. Splitting `DGData`
 
-The number of nodes is computed as `max(node_ids) + 1`. If the `DGraph` is empty, `start_time` and `end_time` are `None`.
+After loading your data, you'll probably want to split your dataset into *train*, *validation*, and *test* splits. TGM provides a **strategy pattern** interface for different split strategies:
+
+- `TemporalSplit`: Split by fixed timestamp boundaries
+- `TemporalRatioSplit`: Split by ratio of events
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md]: split by both edge and node events

Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md]: potentially we can have temporal edge ratio split or temporal node ratio split strategies later on as they might be helpful. 

-- edge ratio split is intuitive for link level tasks
-- node ratio split is intuitive for node level tasks

File: docs/tutorials/dgraph_tutorial.md [hunk 7]
```diff
@@ -111,25 +174,151 @@ dg = DGraph.from_pandas(
...
+# If you tried to override the split strategy, you'll get an error
+
+from tgm.split import TemporalRatioSplit
+split_strategy = TemporalRatioSplit(train=0.8, val=0.1, test=0.1)
+_ data.split(strategy=split_strategy) # Raises ValueError
+```
+
+## 5. Discretizing `DGData`
+
+Dynamic graphs can operate in either continuous time (CTDG) or discrete time (DTDG). You can learn mode about this in [this paper](https://arxiv.org/abs/2407.12269).
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md]: In TGM, we do not enforce strict definition of continuous time and discretize time dynamic graphs, instead, as you have seen, we define the graphs based on their time granularity. Therefore, the user is able to convert between event-based and snapshot-based view of the underlying graph. You can learn mode about this in [the UTG paper](https://arxiv.org/abs/2407.12269)

Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md]: In this way, ML models that are both snapshot and event based can be applied on the same network depending on how you define the time granularity and how you iterate over data in chronological order.

File: docs/tutorials/dgraph_tutorial.md [hunk 8]
```diff
@@ -139,67 +328,120 @@ print(f'Total events (edge+node)  : {dg.num_events}') # 6
 print(f'Edge feature dimension    : {dg.edge_feats_dim}') # 5
 print(f'Static node feature dim   : {dg.static_node_feats_dim}') # 11
 print(f'Dynamic node feature dim  : {dg.dynamic_node_feats_dim}') # 5
-print('==========================')
+
+print(f'TimeDelta                 : {dg.time_delta}') # TimeDelta('s', value=1)
+print(f'Device                    : {dg.device}') # torch.device(cpu)
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:335]: the device of a `DGraph` determines the device of the `DGBatch` objects when iterating over the dataloader.

File: docs/tutorials/dgraph_tutorial.md (hunk 1 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md:20]: Hmm, yeah, though it has a few methods on it too. I think it's fine to leave so that conceptually users think about `DGBatch` is a group of tensors, in contrast

File: docs/tutorials/dgraph_tutorial.md (hunk 3 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md]: Yes, tehnically, but whether you install from source or from pip you still need the wheels in your python env right?

File: docs/tutorials/dgraph_tutorial.md (hunk 4 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md:91]: Can you elaborate?

File: docs/tutorials/dgraph_tutorial.md (hunk 5 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md:301]: Can you explain how it applies here? I already mention that `DGData` sorts edge and node events chronologically above?

File: docs/tutorials/dgraph_tutorial.md (hunk 6 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md]: Sounds good

File: docs/tutorials/dgraph_tutorial.md (hunk 7 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md]: Good suggestions, adding it.

File: docs/tutorials/dgraph_tutorial.md (hunk 8 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md:335]: Good point, added

File: docs/tutorials/dgraph_tutorial.md (hunk 1 above)
Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md:20]: okay, sounds good.

File: docs/tutorials/dgraph_tutorial.md (hunk 4 above)
Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md:91]: it is basically a way for user to see if the `DGData` has edge events, dynamic node events, static node feature etc. or not, might be helpful for native datasets since the user didn't input it. Or alternatively, they just check on `DGraph`

File: docs/tutorials/dgraph_tutorial.md (hunk 3 above)
Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md]: yes

File: docs/tutorials/dgraph_tutorial.md (hunk 4 above)
Comment (shenyangHuang) [reply docs/tutorials/dgraph_tutorial.md:91]: we just also add a `print(data)` here as well
